### ISSUE TITLE
MXbox -- a simple and flexible vision toolbox for mxnet framework.

### ISSUE THREAD

**Issue by Lyken17** [apache/incubator-mxnet#7376]
Thanks for DMLC generous sharing, we can use mxnet -- a great framework efficient  in both memory and speed. However, after using mxnet for months, I found mxnet still has some inconvenience in developing.

For example, mxnet recommends using `im2rec` for loading dataset, but sometimes we need to handle a dataset where new data comes every day. Then generating `im2rec` becomes a time consuming job since incremental adding is not supported. Also, `rec` seems to originally designed for classification tasks, when handling other tasks like semantic segmentation and nlp machine translation, designing a correct `rec` with proper argumentations can be difficult. So I hope there to be a simple and efficient data loader for various tasks.

Another thing is pretrained models. No matter you are a researcher or engineer, a good baseline is always a good start. Unfortunately, many state-of-art models are not originally implemented in mxnet. And their definition files and pretrained weights are scatted among sites. Let alone sometimes they do not provide a detailed description of the important parameters such as normalization mean and std. It would be great if there is a model zoo providing popular networks together with their pretrained weights and hyper parameters. 

Inspired by [pytorch](https://github.com/pytorch/pytorch) and [torchvision](https://github.com/pytorch/vision),  which give a great flexibility to developing, I write [mxbox](https://github.com/Lyken17/mxbox), -- a simple and flexible vision toolbox for mxnet framework. You can download it by 
```bash
pip install mxbox
```

Now [mxbox](https://github.com/Lyken17/mxbox) provides following features

1. Define **preprocess** as a flow (several backends are supported, so no longer need to worry about opencv2 / 3 compatibility)

```python
transform = transforms.Compose([
    transforms.RandomSizedCrop(224),
    transforms.RandomHorizontalFlip(),
    transforms.mx.ToNdArray(),
    transforms.mx.Normalize(mean = [123.68,116.779,103.939 ]),
])
```

2) Build an efficient multi-thread **DataLoader** in few lines

Common datasets such as `cifar10`, `cifar100`, `SVHN`, `MNIST` are out-of-the-box. You can simply load them from `mxbox.datasets`.

```python
from mxbox import transforms, datasets, DataLoader
trans = transforms.Compose([
        transforms.mx.ToNdArray(), 
        transforms.mx.Normalize(mean = [ 0.485, 0.456, 0.406 ],
                                std  = [ 0.229, 0.224, 0.225 ]),
])
dataset = datasets.CIFAR10('~/.mxbox/cifar10', transform=trans, download=True)

batch_size = 32
feedin_shapes = {
    'batch_size': batch_size,
    'data': [mx.io.DataDesc(name='data', shape=(batch_size, 3, 32, 32), layout='NCHW')],
    'label': [mx.io.DataDesc(name='softmax_label', shape=(batch_size, ), layout='N')]
}
loader = DataLoader(dataset, feedin_shapes, threads=8, shuffle=True)
```  
3) Load **popular model** with pretrained weights
```python
alexnet = mxbox.models.alexnet(num_classes=1000, pretrained=True)
vgg16 = mxbox.models.vgg16(num_classes=1000, pretrained=True)
resnet_diy = mxbox.models.resnet(num_classes=1000, num_units=[6,14,32,64], pretrained=True)
densenet169 = mxbox.models.densenet169(num_classes=1000, pretrained=True)
```

Now most parts have finished construction except `models`. Models provided in [mxnet model zoo](https://github.com/dmlc/mxnet-model-gallery) do not come with definition files (some missing its normalization parameters). And recent popular models like [densenet](https://github.com/liuzhuang13/DenseNet) are not included. I would appreciate if someone can help on this.

Give me a star if you like this project. Any questions, comments or suggestions are welcome!

**Comment by piiswrong** [comment]
Thanks a lot for the contribution. This looks pretty similar to the recently introduced gluon utilities.

see gluon.data and gluon.model_zoo. Would you be interested in merging it into gluon?

**Comment by Lyken17** [comment]
I am going to apply for graduate schools this fall, so the time I can spend on mxbox will be very limited. I think moving mxbox to community would be a good idea. And also, as for pretrained models, I do need help from community. 

But I am thinking, is it better to merge mxbox into gluon.data or keep it as a standalone tools, like the ps-lite and tensorboard in dmlc. Since it is not strictly binded with gluon/ndarray, keeping it standalone can make iterations of develop faster (without changes to mxnet main branch).

**Comment by szha** [comment]
@Lyken17 nice! It's great to see the contribution. I'd make similar suggestion as Eric for the following reasons.

Since Gluon will be the place where most of the development will happen going forward, tools in Gluon will have the most exposure as well as most support. Also, Gluon abstracts away NDArray and Symbol wherever possible, which means that tools that conform to the standard will be usable in both NDArray and Symbol, thus benefiting all use cases. Having it as standalone means that someone has to merge it to master from time to time, and it can mean more time-consuming process (at least 2 PRs in two places to get a bug fixed and to have users pick it up). Given that your time commitment is limited, I think merging it into Gluon will be the best option to ensure continued maintenance.

**Comment by zhreshold** [comment]
Good intuition and hitting the point, but seems like there's a lot of duplicated function with gluon data loader(https://github.com/apache/incubator-mxnet/tree/master/python/mxnet/gluon/data) and mx.image functions(http://mxnet.io/api/python/image.html).

I agree with @piiswrong  @szha to merge in, otherwise we will have to maintain a bunch of duplicated stuffs, which is a nightmare.

**Comment by Lyken17** [comment]
@szha  Gluon's documentation seems to be under construction.  Could u show some examples of `mx.symbol`?  In fact `mxbox` provides [`mx.transforms.lambda`](https://github.com/Lyken17/mxbox/blob/master/mxbox/transforms/general.py#L140)  which allows user to easily create extensions . Thus supports for `mx.symbol` should not be a trouble issue.

@zhreshold Gluon and mxbox are both inspired by torchvision, so they have many same functions. What I think is to keep the relation of `mxbox` and `mxnet`  like `pytorch` and `torchvision` -- standalone but strongly connected. The advantage of doing this is people can develop `mxbox` without affecting `mxnet` (push to `mxbox` only), and user can always enjoy the latest update of mxbox from PyPi.

Maintaining a standalone toolbox will not be a nightmare, instead, it will brings faster update -- for example, a new network comes out, I can upload to `mxnet` and give a release immediately, once I get the definition scripts and the pretrained weight. where in mxnet, it requires a reviewer to merge in and extra time for a new release on PyPi.

We all aim for a good toolbox, and as I mentioned, my free time on next semester is very limited. So to merge in or not, actually depends on later maintainers' idea. If you all think merging to be a better choice, I will create a pull request after construction for `model` is finished.

**Comment by Lyken17** [comment]
Hi there, I am merging MXbox to gluon.data.  Which branch should I create pull request to?

**Comment by szha** [comment]
@Lyken17 thanks! Master branch will do. Once you propose the change, we can see the scope and decide how to proceed from there.

**Comment by szha** [comment]
How's the progress?

**Comment by Lyken17** [comment]
@szha 

Hello, I've just created a pull request.

What I did:
1. Add transform utils to gluon.data.
2. Add datasets to gluon.data.vison

What are left:
1. An efficient multi-process loader
I notice there is already one single process loader. `MXBox` has a multi-thread loader, however due to the GIL of python, multi-thread does not help much when loading large image. Meanwhile, fetching large data from queue will cause large overhead in multi-process. The best solution I have seen so far is the pin_memory in pytorch.Tensor -- which passes a reference between processes instead of copying.

2. More dataesets like SVHN / COCO etc

3. New feature `gpuManager`
Without specifying devices manually, this API can automatically find free GPUs for your program. 
```python
# apply 3 `free` gpus on this server
# `Free`  : Memory usage percentage is under 20% 
gpus = gpuManager.allocate( 3, memory_limit=0.2)  # return a list of available gpus 
```
I just added it `MXBox`, how do you think of merging it?

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.
